Cross-signing again [1/3]

uhoreg opened the pull request:
Together with https://github.com/matrix-org/synapse/pull/5726 and https://github.com/matrix-org/synapse/pull/5727, implements the synapse side of https://github.com/matrix-org/matrix-doc/pull/1756 (as of the current version)
fixes https://github.com/matrix-org/synapse/issues/4110

This PR implements uploading cross-signing keys.  #5726 implements uploading signatures.  #5727 implements the federation bits.

uhoreg commented on synapse/handlers/device.py:
```diff
@@ -419,13 +420,153 @@ def notify_device_update(self, user_id, device_ids):
+            "device_list_key", position, users=[from_user_id],
+        )
+
+    @measure_func("device.get_user_ids_changed")
```
Ignore this function -- it seems to have been accidentally copied (maybe from a bad merge) from line ~95.  Will remove in a future patch.

richvdh commented on synapse/storage/end_to_end_keys.py:
```diff
@@ -281,3 +285,145 @@ def delete_e2e_keys_by_device_txn(txn):
+            key_type (str): the type of key that is being set: either 'master'
+                for a master key, 'self_signing' for a self-signing key, or
+                'user_signing' for a user-signing key
+            key (dict): the key data
```
it might be helpful to give a quick example of the shape of data we expect here.

richvdh commented on synapse/storage/end_to_end_keys.py:
```diff
@@ -281,3 +285,145 @@ def delete_e2e_keys_by_device_txn(txn):
+        # since signatures are identified by device ID.  So add an entry to the
+        # device table to make sure that we don't have a collision with device
+        # IDs
+        for v in key["keys"].values():
```
I think I would spell this as `pubkey = next(iter(key["keys"].values()))`, but ymmv.

More helpfully: could you add a comment documenting the shape of `key` to show why this is the right thing to do?

richvdh commented on synapse/storage/end_to_end_keys.py:
```diff
@@ -281,3 +285,145 @@ def delete_e2e_keys_by_device_txn(txn):
+            desc="store_master_key"
+        )
+
+    def set_e2e_cross_signing_key(self, user_id, key_type, key):
```
there seems to be some confusion here about whether it is a *device* signing key or a *cross* signing key.

richvdh commented on synapse/storage/end_to_end_keys.py:
```diff
@@ -281,3 +285,145 @@ def delete_e2e_keys_by_device_txn(txn):
             "delete_e2e_keys_by_device", delete_e2e_keys_by_device_txn
         )
+
+    def _set_e2e_device_signing_key_txn(self, txn, user_id, key_type, key):
```
should this be called `_set_e2e_cross_signing_key_txn` ?

richvdh commented on synapse/storage/end_to_end_keys.py:
```diff
@@ -281,3 +285,145 @@ def delete_e2e_keys_by_device_txn(txn):
+            key_type (str): the type of key that is being set: either 'master'
+                for a master key, 'self_signing' for a self-signing key, or
+                'user_signing' for a user-signing key
+            from_user_id (str): if specified, signatures made by this user on
```
would it be clearer if fetching the signatures was pulled out to a separate function? It would have the advantage of adding symmetry with `store_e2e_device_signatures`.

uhoreg replied:
The difference is that when you fetch a cross-signing key, you always want all the signatures associated with it (that are visible to that user), whereas signatures can be uploaded separately, and so have a separate function for storing them.

richvdh replied:
I still feel like this "convenience" code belongs outside the storage layer, but up to you

uhoreg commented on synapse/storage/schema/delta/55/signing_keys.sql:
```diff
@@ -0,0 +1,48 @@
+    user_ids TEXT NOT NULL
+);
+
+CREATE INDEX user_signature_stream_idx ON user_signature_stream(stream_id, from_user_id);
```
I think I basically cargo-culted the `device_list_stream` from `storage/schema/delta/40/device_list_streams.sql`, which just has an index on `(stream_id, user_id)`.  I don't really understand how stream IDs work, but I can add a unique index on `stream_id`.  `from_user_id` is use in the `WHERE` clause of the query, so it seems to me like it's useful in the index.

richvdh replied:
I wouldn't necessarily take the existing database code as a great example :/.

The query in question is `https://github.com/matrix-org/synapse/commit/8567ff6b4938a477c5b82c5c235124b1ca2e5fbe#diff-21f2f9be6a999959ddb0f9e3769006ccR361` ? Since the `stream_id` part of the query is an open-ended range, it won't be able to make any use of the `user_id` part of the index.

uhoreg commented on synapse/storage/end_to_end_keys.py:
```diff
@@ -85,7 +89,7 @@ def _get_e2e_device_keys_txn(
             " FROM devices d"
             "    %s JOIN e2e_device_keys_json k USING (user_id, device_id)"
-            " WHERE %s"
+            " WHERE (%s) AND d.hidden = 0"
```
`NOT d.hidden` doesn't seem to work if `d.hidden` is `NULL`, AFAICT.  I think it would need to be `AND (NOT d.hidden OR d.hidden ISNULL)`

uhoreg commented on synapse/storage/schema/delta/55/signing_keys.sql:
```diff
@@ -0,0 +1,48 @@
+
+-- device list needs to know which ones are "real" devices, and which ones are
+-- just used to avoid collisions
+ALTER TABLE devices ADD COLUMN hidden BOOLEAN DEFAULT FALSE;
```
I have some queries that look for hidden = false (e.g. https://github.com/matrix-org/synapse/pull/4970/commits/8567ff6b4938a477c5b82c5c235124b1ca2e5fbe#diff-21f2f9be6a999959ddb0f9e3769006ccR53 ).  Is there a way to make them still work if the column is nullable?

richvdh replied:
You probably need custom sql rather than `_simple_select_one`. `NULL` is falsey in both sqlite and postgres so you can do `select * from devices where not hidden`.

richvdh replied:
... no you can't. You need `where hidden is not true`. https://stackoverflow.com/a/46474204/637864

richvdh replied:
... except that doesn't work in sqlite. I hate computers.

`select * from devices where not coalesce(hidden, ?);` and pass `False` as a parameter.

It might be easier to ignore `hidden` in the sql and do the filtering in the python instead.

uhoreg commented:
I think I've addressed all the concerns.

Note that the sytest is set up to run against this PR rather than the last PR, so it fails.  I can fix this by either splitting the sytest PR into three, or changing the sytest PR to run against the last PR (which I think I need to do by creating a new sytest PR.)

richvdh commented on synapse/types.py:
```diff
@@ -475,3 +478,24 @@ class ReadReceipt(object):
+    """
+    # make sure that exactly one key is provided
+    if "keys" not in key_info:
+        raise SynapseError(400, "Invalid key")
```
Isn't this more of a ValueError than a SynapseError? I don't think it's up to a utility function to determine whether this translates into a 400 error at the API or not: that's a question for the REST layer or the handlers.

For example: the key passed into this function might have come out of the database. Where that is the case, the problem isn't with the request, but with the database, so the appropriate response is a 500, not a 400.

richvdh commented on synapse/storage/schema/delta/55/signing_keys.sql:
```diff
@@ -0,0 +1,59 @@
+    user_id TEXT NOT NULL,
+    -- the type of cross-signing key (master, user_signing, or self_signing)
+    keytype TEXT NOT NULL,
+    -- the full key information
```
```suggestion
    -- the full key information, as a json-encoded dict
```

richvdh commented on synapse/storage/schema/delta/55/signing_keys.sql:
```diff
@@ -0,0 +1,59 @@
+    -- the full key information
+    keydata TEXT NOT NULL,
+    -- time that the key was added
+    ts BIGINT NOT NULL
```
ts is kinda hard to grep for and not very descriptive. can you call it `added_ts`, maybe?

richvdh commented on synapse/storage/schema/delta/55/signing_keys.sql:
```diff
@@ -0,0 +1,59 @@
+/* Copyright 2019 New Vector Ltd
```
this is going to have to move to /56/ now, btw

richvdh commented on synapse/handlers/e2e_keys.py:
```diff
@@ -342,6 +440,105 @@ def _upload_one_time_keys_for_user(
+        signing_key (VerifyKey): (optional) the signing key that the key should
+            be signed with.  If omitted, signatures will not be checked.
+    """
+    if (
```
this could probably do with checking the existence and validity of the `keys` property?

richvdh commented on synapse/handlers/e2e_keys.py:
```diff
@@ -342,6 +440,105 @@ def _upload_one_time_keys_for_user(
+            be signed with.  If omitted, signatures will not be checked.
+    """
+    if (
+        "user_id" not in key
```
`key.get("user_id") != user_id` is probably a more succinct way to write these two lines

richvdh commented on synapse/storage/devices.py:
```diff
@@ -552,12 +617,25 @@ def store_device(self, user_id, device_id, initial_device_display_name):
             self.device_id_exists_cache.prefill(key, True)
             defer.returnValue(inserted)
+        except StoreError as e:
+            raise e
```
You want to allow the existing exception to propagate, not rethrow it with a new stack trace:

```suggestion
            raise
```

richvdh commented on synapse/rest/client/v2_alpha/keys.py:
```diff
@@ -227,8 +229,46 @@ def on_POST(self, request):
+    }
+    """
+
+    PATTERNS = client_patterns("/keys/device_signing/upload$")
```
is this meant to be on `r0` or `unstable`? please can it only be on one of them, anyway?

uhoreg replied:
I guess it goes on `unstable` until the MSC goes through?

richvdh replied:
yeah, I think so.

uhoreg commented on synapse/handlers/e2e_keys.py:
```diff
@@ -64,6 +70,11 @@ def query_devices(self, query_body, timeout):
+        Args:
+            from_user_id (str): the user making the query.  This is used when
+                adding cross-signing signatures to limit what signatures users
+                can see.
```
I didn't realize that the test was the only other place that called this function.  I will indeed update the test and drop the default.

richvdh commented:
so I guess this is superceded by #5759 and #5769.
